src/query/cellxgene_census/script.py

```diff
@@ -0,0 +1,194 @@
+        cell_of_interest_terms.extend(
+            list(lower_hierarchy_cell_of_interest_map.keys())
+            )
+
+    logger.info(lower_hierarchy_cell_of_interest_map)
+
+    return lower_hierarchy_cell_of_interest_map, cell_of_interest_terms
+
+
+# TODO: function to explore cell types available in query data
+# def view_available_cell_types(lower_hierarchy_cell_of_interest_map, cell_of_interest_terms):
+#     cells_of_interest_query_terms = list(adata.obs.cell_type_ontology_term_id.unique())
+#     avail_cells_overview = []
+#     for term in cell_of_interest_terms:
+#         if term in cells_of_interest_query_terms:
+#             avail_cells_overview.append("Available: {}-{}".format(term, lower_hierarchy_cell_of_interest_map[term]))
+#         else:
+#             avail_cells_overview.append("Unavailable: {}-{}".format(term, lower_hierarchy_cell_of_interest_map[term]))
+#     avail_cells_overview.sort()
+#     logging.info(avail_cells_overview)
+
+
+def build_census_query():
+    _query = f'is_primary_data == {par["is_primary_data"]}'
+    if par["cell_type"]:
+        _, cell_of_interest_terms = get_cell_type_terms()
+        _query = _query + f' and cell_type_ontology_term_id in {cell_of_interest_terms}'
+    if par["tissue"]:
+        _query = _query + f' and tissue in {par["tissue"]}'
+    if par["technology"]:
+        _query = _query + f' and assay in {par["technology"]}'
+    if par["suspension"]:
+        _query = _query + f' and suspension_type in {par["suspension"]}'
+    logger.info("Census query: %s", _query)
+
+    return _query
+
+
+def extract_metadata(census_connection, query):
+    logger.info("Extracting only metadata")
+    logger.info(census_connection["census_data"][par["species"]].obs)
+    return (
+        census_connection["census_data"][par["species"]].obs.read(
+            value_filter=query,
+            column_names=par["obs_column_names"]).concat().to_pandas()
+        )
+
+
+def extract_metadata_expression(census_connection, query):
+    logger.info("Extracting metadata and gene expression matrix")
+    return (
+        cellxgene_census.get_anndata(
+            census_connection,
+            organism = par["species"],
+            obs_value_filter = query,
+            column_names = {"obs": par["obs_column_names"]}
+            )
+        )
+
+
+def write_mudata(mdata):
+    logger.info("Writing %s", par["output"])
+    mdata.write_h5mu(
+        par["output"],
+        compression = par["output_compression"]
+        )
+
+
+def main():
+
+    census_connection = connect_census()
+    query = build_census_query()
+
+    if par["metadata_only"]:
+        query_data = extract_metadata(census_connection, query)
+
+    else:
+        query_data = extract_metadata_expression(census_connection, query)
+
+    # clean-up mem
+    census_connection.close()
+    del census_connection
+
+    if isinstance(query_data, pd.core.frame.DataFrame):
```

Comment:
I would let both `extract_metadata` and `extract_metadata_expression` return a ad.AnnData object and move the `csr_matrix(query_data.X)` into the `extract_metadata_expression`. This can avoid duplication of 

```
 mdata = mu.MuData(
      {par["modality"]: query_data}
      )
```

Reply:
Excellent idea!